lib/chef/provisioning/aws_driver/driver.rb

```diff
@@ -101,23 +115,83 @@ def allocate_load_balancer(action_handler, lb_spec, lb_options, machine_specs)
-        enable_zones = (availability_zones || []).dup
-        disable_zones = []
-        actual_elb.availability_zones.each do |availability_zone|
-          if !enable_zones.delete(availability_zone.name)
-            disable_zones << availability_zone.name
+        # TODO: refactor this whole giant method into many smaller method calls
+        # Update scheme - scheme is immutable once set, so if it is changing we need to delete the old
+        # ELB and create a new one
+        if scheme.downcase != actual_elb.scheme
+          desc = ["  updating scheme to #{scheme}"]
+          desc << "  WARN: scheme is immutable, so deleting and re-creating the ELB"
+          perform_action.call(desc) do
+            actual_elb.delete
+            actual_elb = elb.load_balancers.create(lb_spec.name, lb_optionals)
           end
         end
-        if enable_zones.size > 0
-          perform_action.call("  enable availability zones #{enable_zones.join(', ')}") do
-            actual_elb.availability_zones.enable(*enable_zones)
+
+        # Update security groups
+        if security_group_ids.empty?
+            Chef::Log.info("No Security Groups specified in resource; load_balancer[#{actual_elb.name}] cannot have " +
+            "empty Security Groups, so assuming it only currently has the default Security Group.  No action taken.")
+        else
+          current = Set.new(actual_elb.security_group_ids)
+          desired = Set.new(security_group_ids)
+          if current != desired
+            perform_action.call("  updating security groups to #{desired.to_a}") do
+              elb.client.apply_security_groups_to_load_balancer(
+                load_balancer_name: actual_elb.name,
+                security_groups: desired.to_a
+              )
+            end
           end
         end
-        if disable_zones.size > 0
-          perform_action.call("  disable availability zones #{disable_zones.join(', ')}") do
-            actual_elb.availability_zones.disable(*disable_zones)
+
+        # You can only have 1 of either subnets or availability zones, and if in a VPC you must have
+        # at least one of either.
+        if !availability_zones.empty?
+
+          enable_zones = (availability_zones || []).dup
+          disable_zones = []
+          actual_elb.availability_zones.each do |availability_zone|
+            if !enable_zones.delete(availability_zone.name)
+              disable_zones << availability_zone.name
+            end
+          end
+          if enable_zones.size > 0
+            perform_action.call("  enable availability zones #{enable_zones.join(', ')}") do
+              actual_elb.availability_zones.enable(*enable_zones)
+            end
           end
+          if disable_zones.size > 0
+            perform_action.call("  disable availability zones #{disable_zones.join(', ')}") do
+              actual_elb.availability_zones.disable(*disable_zones)
+            end
+          end
+
+        elsif !subnets.empty?
```

Comment:
My concern is this: if you start with subnets and switch to availability zones, then it will just add the availability zones, when it *should* delete the subnets as well.  More specifically, I think this case will have an issue:

First converge:
```ruby
load_balancer 'blah' do
  availability_zones [ 'a', 'b', 'c' ]
end
```

Second converge:

```ruby
load_balancer 'blah' do
  subnets [ 'd', 'e', 'f' ]
end
```

I would expect the second recipe to remove the availability zones as well as add the subnets, but I don't think your code will do the removing.

At least, that's how the code reads to me.  I think you want an algorithm more like this:

- Delete any subnets that the user didn't ask for
- Delete any availability zones that the user didn't ask for
- Add any subnets the user asked for (and don't already exist)
- Add any availability zones the user asked for (and don't already exist)

Where the current flow is more like:

- If the user specified availability zones,
  - Delete any availability zones that the user didn't ask for
  - Add any availability zones the user asked for (and don't already exist)
- If the user specified subnets and did NOT specify availability zones,
  - Delete any subnets the user didn't ask for
  - Add any subnets the user asked for (and don't already exist)

Bonus is fewer branches in the logic means more of the logic is part of the normal flow (and therefore tested at least a little bit every time you run).